flashbax/vault/vault.py

```diff
@@ -0,0 +1,448 @@
+    different behaviour for dictionaries (DictKey) vs. namedtuples (GetAttrKey), which means
+    we could not save a vault based on a namedtuple structure but later load it as a dict.
+    Instead, this function maps both to a consistent string representation.
+
+    Args:
+        path: tuple of DictKeys or GetAttrKeys
+
+    Returns:
+        str: standardised string representation of the path
+    """
+    path_str = ""
+    for p in path:
+        if isinstance(p, DictKey):
+            path_str += str(p.key)
+        elif isinstance(p, GetAttrKey):
+            path_str += p.name
+        path_str += "."
+    return path_str
+
+
+class Vault:
+    def __init__(
+        self,
+        vault_name: str,
+        experience_structure: Optional[Experience] = None,
+        rel_dir: str = "vaults",
+        vault_uid: Optional[str] = None,
+        metadata: Optional[dict] = None,
+    ) -> None:
+        """Flashbax utility for storing buffers to disk efficiently.
+
+        Args:
+            vault_name (str): the upper-level name of this vault.
+                Resulting path is <cwd/rel_dir/vault_name/vault_uid>.
+            experience_structure (Optional[Experience], optional):
+                Structure of the experience data, usually given as `buffer_state.experience`.
+                Defaults to None, which can only be done if reading an existing vault.
+            rel_dir (str, optional):
+                Base directory of all vaults. Defaults to "vaults".
+            vault_uid (Optional[str], optional): Unique identifier for this vault.
+                Defaults to None, which will use the current timestamp.
+            metadata (Optional[dict], optional):
+                Any additional metadata to save. Defaults to None.
+
+        Raises:
+            ValueError:
+                If the targeted vault does not exist, and no experience_structure is provided.
+
+        Returns:
+            Vault: a vault object.
+        """
+        # Get the base path for the vault and the metadata path
+        vault_str = vault_uid if vault_uid else datetime.now().strftime("%Y%m%d%H%M%S")
+        self._base_path = os.path.join(os.getcwd(), rel_dir, vault_name, vault_str)
+        metadata_path = epath.Path(os.path.join(self._base_path, METADATA_FILE))
+
+        # Check if the vault exists, otherwise create the necessary dirs and files
+        base_path_exists = os.path.exists(self._base_path)
+        if base_path_exists:
+            print(f"Loading vault found at {self._base_path}")
+            # Vault exists, so we load the metadata to access the structure etc.
+            self._metadata = json.loads(metadata_path.read_text())
+
+            # Ensure minor versions match
+            assert (self._metadata["version"] // 1) == (VERSION // 1)
+
+        elif experience_structure is not None:
+            print(f"New vault created at {self._base_path}")
```

Comment:
Can we put this at the bottom, just in case something within this block fails, you don't get the print out and an error

Reply:
Done 👍🏻